flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/PartitionableTableSource.scala

```diff
@@ -0,0 +1,193 @@
+    prunedPartitions: JList[Partition],
+    predicates: JList[Expression]): TableSource[T]
+
+
+  /**
+    * Check and pick all predicates this table source can support. The passed in predicates
+    * have been translated in conjunctive form, and table source can only pick those predicates
+    * that it supports.
+    * <p>
+    * After trying to push predicates down, we should return a new [[TableSource]]
+    * instance which holds all pushed down predicates. Even if we actually pushed nothing down,
+    * it is recommended that we still return a new [[TableSource]] instance since we will
+    * mark the returned instance as filter push down has been tried.
+    * <p>
+    * We also should note to not changing the form of the predicates passed in. It has been
+    * organized in CNF conjunctive form, and we should only take or leave each element from the
+    * list. Don't try to reorganize the predicates if you are absolutely confident with that.
+    *
+    * @param predicates A list contains conjunctive predicates, you should pick and remove all
+    *                   expressions that can be pushed down. The remaining elements of this list
+    *                   will further evaluated by framework.
+    * @return A new cloned instance of [[TableSource]].
+    */
+  override def applyPredicate(predicates: JList[Expression]): TableSource[T] = {
+    var partitionPruned = false
+    var prunedPartitions: JList[Partition] = new JArrayList()
+
+    // extract partition predicate
+    val (partitionPredicates, _) = PartitionPredicateExtractor.extractPartitionPredicates(
+      predicates.asScala.toArray, getPartitionFieldNames)
+    if (partitionPredicates.nonEmpty) {
+      // do partition pruning
+      val builder = relBuilder.getOrElse(throw new TableException("relBuilder is null"))
+      prunedPartitions = applyPartitionPruning(partitionPredicates, builder)
+      partitionPruned = true
+    }
+
+    if (supportDropPartitionPredicate) {
+      predicates.removeAll(partitionPredicates.toList.asJava)
+    }
+
+    applyPrunedPartitionsAndPredicate(partitionPruned, prunedPartitions, predicates)
+  }
+
+  /**
+    * @param relBuilder Builder for relational expressions.
+    */
+  def setRelBuilder(relBuilder: RelBuilder): Unit = {
+    this.relBuilder = Some(relBuilder)
+  }
+
+  /**
+    * Default implementation for partition pruning.
+    *
+    * @param partitionPredicates A filter expression that will be applied against partition values.
+    * @param relBuilder          Builder for relational expressions.
+    * @return The pruned partitions.
+    */
+  def applyPartitionPruning(
+    partitionPredicates: Array[Expression],
+    relBuilder: RelBuilder): JList[Partition] = {
+    PartitionPruner.INSTANCE.getPrunedPartitions(
+      getPartitionFieldNames,
+      getPartitionFieldTypes,
+      getAllPartitions,
+      partitionPredicates,
+      relBuilder)
+  }
+
+}
+
+/**
+  * The base class of partition
+  */
+trait Partition {
```

Comment:
Can you provide more detailed description about what is a "Partition" and how one PartitionableTableSource will do partition pruning. User cannot get precise intuition about what is a field of partition, and what does origin value mean.

Reply:
I will add more description about them.

The origin value means the entire partition value in the `Partition` instance. A partition value may be simple, such as the data is splited by year (year=2015,  year=2016); and  A partition value may be complex, such as the data is splitted by year and month (year=2015,month=01,  year=2015,month=02,  year=2016,month=01,   year=2016,month=02).